core/network/source_netlink.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package network
+
+import (
+	"net"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+	"github.com/vishvananda/netlink"
+)
+
+// netlinkAddr implements ConfigSourceAddr based on the
+// netlink implementation of a network address.
+type netlinkAddr struct {
+	addr *netlink.Addr
+}
+
+// IP (ConfigSourceAddr) is a simple property accessor.
+func (a *netlinkAddr) IP() net.IP {
+	return a.addr.IP
+}
+
+// IPNet (ConfigSourceAddr) is a simple property accessor.
+func (a *netlinkAddr) IPNet() *net.IPNet {
+	return a.addr.IPNet
+}
+
+// String (ConfigSourceAddr) is a simple property accessor.
+func (a *netlinkAddr) String() string {
+	return a.addr.String()
+}
+
+// netlinkNIC implements ConfigSourceNIC by wrapping a netlink Link.
+type netlinkNIC struct {
+	nic      netlink.Link
+	getAddrs func(netlink.Link) ([]netlink.Addr, error)
+}
+
+// Name returns the name of the device.
+func (n netlinkNIC) Name() string {
+	return n.nic.Attrs().Name
+}
+
+// Type returns the interface type of the device.
+func (n netlinkNIC) Type() InterfaceType {
+	switch n.nic.Type() {
+	case "bridge":
+		return BridgeInterface
+	case "vlan":
+		return VLAN_8021QInterface
+	case "bond":
+		return BondInterface
+	}
+
+	if n.nic.Attrs().Flags&net.FlagLoopback > 0 {
+		return LoopbackInterface
+	}
+
+	return EthernetInterface
```

Comment:
Can we add a comment here similar to the interface one, that this should be revised accordingly to show that this isn't always going to be correct for veth, tuntap, macvtap etc.